SignXML Verifier Fails When XML Signature Uses Default Namespace Without ds: Prefix

I'm working on generating a signed XML document where the **<Signature>** element (and related elements like **<SignedInfo>**, **<DigestMethod>**, etc.) must not include the **ds:** prefix. This is because the legacy application I’m integrating with expects the XML Digital Signature block to be present without any namespace prefixes—only the Signature tag and children in the correct namespace (i.e., http://www.w3.org/2000/09/xmldsig#), but without the ds: prefix.

I’m using signxml to generate the signature. Signing works fine when I override the namespace as:

```python
signer.namespaces = {None: signxml.namespaces.ds}
```

This gives me the desired XML structure (no **ds:**), and the output looks valid. However, when I try to verify the signature using `signxml.XMLVerifier().verify(...)`, it fails. The same test passes if I don't override the namespace and let SignXML add the ds: prefix.

#### Test Result
```text
self = <signxml.verifier.XMLVerifier object at 0xffffadce0c50>, element = <Element {http://www.w3.org/2000/09/xmldsig#}Reference at 0xffffacd2e9c0>, query = 'DigestMethod'
require = True, xpath = ''

    def _find(self, element, query, require=True, xpath=""):
        namespace = "ds"
        if ":" in query:
            namespace, _, query = query.partition(":")
        result = element.find(f"{xpath}{namespace}:{query}", namespaces=namespaces)
    
        if require and result is None:
>           raise InvalidInput(f"Expected to find XML element {query} in {element.tag}")
E           signxml.exceptions.InvalidInput: Expected to find XML element DigestMethod in {http://www.w3.org/2000/09/xmldsig#}Reference

/py/lib/python3.12/site-packages/signxml/processor.py:98: InvalidInput
================================================================================ short test summary info ================================================================================
FAILED test_signxml.py::test_signed_without_ns_success - signxml.exceptions.InvalidInput: Expected to find XML element DigestMethod in {http://www.w3.org/2000/09/xmldsig#}Reference
============================================================================== 1 failed, 1 passed in 0.21s ==============================================================================

```

#### Test Code

```python
"""Test Certificate SetUp

.. code-block:: bash

openssl genrsa -out private.key 4096
openssl req -new -key private.key -out request.csr
openssl x509 -req -days 365 -in request.csr -signkey private.key -out certificate.crt


"""

from xml.etree.ElementTree import Element, SubElement, tostring

import lxml
import signxml
from cryptography import x509
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import serialization

PRIVATE_KEY_PATH = "/app/private.key"
CERTIFICATE_PATH = "/app/certificate.crt"


def load_cert_pub(cert_path: str):
    with open(cert_path, "rb") as cert_file:
        cert_data = cert_file.read()
    cert = x509.load_pem_x509_certificate(cert_data, default_backend())
    return cert, cert.public_key()


def load_private_key(key_path: str, password=None):
    with open(key_path, "rb") as key_file:
        key_data = key_file.read()
    key = serialization.load_pem_private_key(
        key_data, password=password, backend=default_backend()
    )
    return key


def create_sample_xml():
    root = Element("ROOT")

    # Section Data
    data_sec = SubElement(root, "DATA")

    pid_elem = SubElement(data_sec, "ID")
    pid_elem.text = "encrypted_some_data_data"

    session_key_elem = SubElement(data_sec, "SESSION_KEY")
    session_key_elem.text = "encrypted_session_key"

    return tostring(root, encoding="unicode")


private_key = load_private_key(PRIVATE_KEY_PATH)
cert, public_key = load_cert_pub(CERTIFICATE_PATH)


def test_signed_with_default_ns_success():
    xml_str = create_sample_xml()
    root = lxml.etree.fromstring(xml_str.encode("utf-8"))

    signer = signxml.XMLSigner(
        method=signxml.methods.enveloped,
        signature_algorithm="rsa-sha256",
        digest_algorithm="sha256",
        c14n_algorithm="http://www.w3.org/TR/2001/REC-xml-c14n-20010315",
    )

    signed_root = signer.sign(
        root,
        key=private_key,
        exclude_c14n_transform_element=True,
    )

    signed_xml_str = lxml.etree.tostring(signed_root, encoding="utf-8")

    # Verifying
    root = lxml.etree.fromstring(signed_xml_str)

    result = signxml.XMLVerifier().verify(root, x509_cert=cert)

    # No exceptions, so this is okay...


def test_signed_without_ns_success():
    # This test is failing...
    xml_str = create_sample_xml()
    root = lxml.etree.fromstring(xml_str.encode("utf-8"))

    signer = signxml.XMLSigner(
        method=signxml.methods.enveloped,
        signature_algorithm="rsa-sha256",
        digest_algorithm="sha256",
        c14n_algorithm="http://www.w3.org/TR/2001/REC-xml-c14n-20010315",
    )

    signer.namespaces = {None: signxml.namespaces.ds}
    # signer.signature_annotators = []  # remove key info from payload

    signed_root = signer.sign(
        root,
        key=private_key,
        exclude_c14n_transform_element=True,
    )

    signed_xml_str = lxml.etree.tostring(signed_root, encoding="utf-8")

    # Verifying
    root = lxml.etree.fromstring(signed_xml_str)
    # This going to raise:
    # signxml.exceptions.InvalidInput: Expected to find XML element DigestMethod in {http://www.w3.org/2000/09/xmldsig#}Reference
    result = signxml.XMLVerifier().verify(root, x509_cert=cert)

```

---

> Side Note:
>I’m new to XML digital signatures, and trying to get this working based on the requirements of the external (legacy) application. If I’m missing something or doing in a wrong way, I'd appreciate your help on the correct approach.

--- a/signxml/signer.py
+++ b/signxml/signer.py
@@ -7,7 +7,7 @@ from cryptography.hazmat.primitives.asymmetric import dsa, ec, rsa, utils
 from cryptography.hazmat.primitives.asymmetric.padding import MGF1, PSS, PKCS1v15
 from cryptography.hazmat.primitives.hmac import HMAC
 from cryptography.hazmat.primitives.serialization import Encoding, load_pem_private_key
-from lxml.etree import Element, SubElement, _Element
+from lxml.etree import Element, QName, SubElement, _Element
 
 from .algorithms import (
     CanonicalizationMethod,
@@ -116,6 +116,20 @@ class XMLSigner(XMLSignatureProcessor):
         self.namespaces = dict(ds=namespaces.ds)
         self._parser = None
         self.signature_annotators = [self._add_key_info]
+
+    def _ds_tag(self, tag):
+        """
+        Create a QName for the ds namespace, respecting the configured namespace mapping.
+
+        When the default namespace is set to the ds namespace ({None: namespaces.ds}),
+        elements should be created without an explicit namespace so they inherit from
+        the nsmap context. This avoids spurious xmlns="" undeclarations in C14N output.
+
+        See https://github.com/XML-Security/signxml/issues/275
+        """
+        if None in self.namespaces and self.namespaces[None] == namespaces.ds:
+            return QName(None, tag)
+        return ds_tag(tag)
 
     def check_deprecated_methods(self):
         if "SHA1" in self.sign_alg.name or "SHA1" in self.digest_alg.name:
@@ -304,9 +318,9 @@ class XMLSigner(XMLSignatureProcessor):
         if self.sign_alg.name.startswith("HMAC_"):
             return
         if signing_settings.key_info is None:
-            key_info = SubElement(sig_root, ds_tag("KeyInfo"))
+            key_info = SubElement(sig_root, self._ds_tag("KeyInfo"))
             if signing_settings.key_name is not None:
-                keyname = SubElement(key_info, ds_tag("KeyName"))
+                keyname = SubElement(key_info, self._ds_tag("KeyName"))
                 keyname.text = signing_settings.key_name
 
             if signing_settings.cert_chain is None or signing_settings.always_add_key_value:
@@ -314,9 +328,9 @@ class XMLSigner(XMLSignatureProcessor):
 
             if signing_settings.cert_chain is not None:
                 assert len(signing_settings.cert_chain) > 0
-                x509_data = SubElement(key_info, ds_tag("X509Data"))
+                x509_data = SubElement(key_info, self._ds_tag("X509Data"))
                 for cert in signing_settings.cert_chain:
-                    x509_certificate = SubElement(x509_data, ds_tag("X509Certificate"))
+                    x509_certificate = SubElement(x509_data, self._ds_tag("X509Certificate"))
                     if isinstance(cert, (str, bytes)):
                         x509_certificate.text = strip_pem_header(cert)
                     else:
@@ -333,7 +347,7 @@ class XMLSigner(XMLSignatureProcessor):
         return c14n_inputs, new_references
 
     def _unpack(self, data, references: List[SignatureReference]):
-        sig_root = Element(ds_tag("Signature"), nsmap=self.namespaces)
+        sig_root = Element(self._ds_tag("Signature"), nsmap=self.namespaces)
         if self.construction_method == SignatureConstructionMethod.enveloped:
             if isinstance(data, (str, bytes)):
                 raise InvalidInput("When using enveloped signature, **data** must be an XML element")
@@ -376,7 +390,7 @@ class XMLSigner(XMLSignatureProcessor):
                 c14n_inputs = [self.get_root(data)]
         elif self.construction_method == SignatureConstructionMethod.enveloping:
             doc_root = sig_root
-            c14n_inputs = [Element(ds_tag("Object"), nsmap=self.namespaces, Id="object")]
+            c14n_inputs = [Element(self._ds_tag("Object"), nsmap=self.namespaces, Id="object")]
             if isinstance(data, (str, bytes)):
                 c14n_inputs[0].text = data
             else:
@@ -389,14 +403,14 @@ class XMLSigner(XMLSignatureProcessor):
     ):
         assert reference.c14n_method is not None
         if self.construction_method == SignatureConstructionMethod.enveloped:
-            SubElement(transforms_node, ds_tag("Transform"), Algorithm=SignatureConstructionMethod.enveloped.value)
+            SubElement(transforms_node, self._ds_tag("Transform"), Algorithm=SignatureConstructionMethod.enveloped.value)
             if not exclude_c14n_transform_element:
-                SubElement(transforms_node, ds_tag("Transform"), Algorithm=reference.c14n_method.value)
+                SubElement(transforms_node, self._ds_tag("Transform"), Algorithm=reference.c14n_method.value)
         else:
             if not exclude_c14n_transform_element:
                 c14n_xform = SubElement(
                     transforms_node,
-                    ds_tag("Transform"),
+                    self._ds_tag("Transform"),
                     Algorithm=reference.c14n_method.value,
                 )
             if reference.inclusive_ns_prefixes:
@@ -407,41 +421,41 @@ class XMLSigner(XMLSignatureProcessor):
     def _build_sig(
         self, sig_root, references, c14n_inputs, inclusive_ns_prefixes, exclude_c14n_transform_element=False
     ):
-        signed_info = SubElement(sig_root, ds_tag("SignedInfo"), nsmap=self.namespaces)
-        sig_c14n_method = SubElement(signed_info, ds_tag("CanonicalizationMethod"), Algorithm=self.c14n_alg.value)
+        signed_info = SubElement(sig_root, self._ds_tag("SignedInfo"), nsmap=self.namespaces)
+        sig_c14n_method = SubElement(signed_info, self._ds_tag("CanonicalizationMethod"), Algorithm=self.c14n_alg.value)
         if inclusive_ns_prefixes:
             SubElement(sig_c14n_method, ec_tag("InclusiveNamespaces"), PrefixList=" ".join(inclusive_ns_prefixes))
 
-        SubElement(signed_info, ds_tag("SignatureMethod"), Algorithm=self.sign_alg.value)
+        SubElement(signed_info, self._ds_tag("SignatureMethod"), Algorithm=self.sign_alg.value)
         for i, reference in enumerate(references):
             if reference.c14n_method is None:
                 reference = replace(reference, c14n_method=self.c14n_alg)
             if reference.inclusive_ns_prefixes is None:
                 reference = replace(reference, inclusive_ns_prefixes=inclusive_ns_prefixes)
-            reference_node = SubElement(signed_info, ds_tag("Reference"), URI=reference.URI)
-            transforms = SubElement(reference_node, ds_tag("Transforms"))
+            reference_node = SubElement(signed_info, self._ds_tag("Reference"), URI=reference.URI)
+            transforms = SubElement(reference_node, self._ds_tag("Transforms"))
             self._build_transforms_for_reference(
                 transforms_node=transforms,
                 reference=reference,
                 exclude_c14n_transform_element=exclude_c14n_transform_element,
             )
-            SubElement(reference_node, ds_tag("DigestMethod"), Algorithm=self.digest_alg.value)
-            digest_value = SubElement(reference_node, ds_tag("DigestValue"))
+            SubElement(reference_node, self._ds_tag("DigestMethod"), Algorithm=self.digest_alg.value)
+            digest_value = SubElement(reference_node, self._ds_tag("DigestValue"))
             payload_c14n = self._c14n(
                 c14n_inputs[i], algorithm=reference.c14n_method, inclusive_ns_prefixes=reference.inclusive_ns_prefixes
             )
             digest = self._get_digest(payload_c14n, algorithm=self.digest_alg)
             digest_value.text = b64encode(digest).decode()
-        signature_value = SubElement(sig_root, ds_tag("SignatureValue"))
+        signature_value = SubElement(sig_root, self._ds_tag("SignatureValue"))
         return signed_info, signature_value
 
     def _build_signature_properties(self, signature_properties):
         # FIXME: make this use the annotator API
-        obj = Element(ds_tag("Object"), attrib={"Id": "prop"}, nsmap=self.namespaces)
-        signature_properties_el = Element(ds_tag("SignatureProperties"))
+        obj = Element(self._ds_tag("Object"), attrib={"Id": "prop"}, nsmap=self.namespaces)
+        signature_properties_el = Element(self._ds_tag("SignatureProperties"))
         for i, el in enumerate(signature_properties):
             signature_property = Element(
-                ds_tag("SignatureProperty"),
+                self._ds_tag("SignatureProperty"),
                 attrib={
                     "Id": el.attrib.pop("Id", f"sigprop{i}"),
                     "Target": el.attrib.pop("Target", f"#sigproptarget{i}"),
@@ -456,17 +470,17 @@ class XMLSigner(XMLSignatureProcessor):
         """
         Add the public components of the key to the signature (see https://www.w3.org/TR/xmldsig-core2/#sec-KeyValue).
         """
-        key_value = SubElement(key_info_node, ds_tag("KeyValue"))
+        key_value = SubElement(key_info_node, self._ds_tag("KeyValue"))
         if self.sign_alg.name.startswith("RSA_") or self.sign_alg.name.startswith("SHA"):
-            rsa_key_value = SubElement(key_value, ds_tag("RSAKeyValue"))
-            modulus = SubElement(rsa_key_value, ds_tag("Modulus"))
+            rsa_key_value = SubElement(key_value, self._ds_tag("RSAKeyValue"))
+            modulus = SubElement(rsa_key_value, self._ds_tag("Modulus"))
             modulus.text = b64encode(long_to_bytes(key.public_key().public_numbers().n)).decode()
-            exponent = SubElement(rsa_key_value, ds_tag("Exponent"))
+            exponent = SubElement(rsa_key_value, self._ds_tag("Exponent"))
             exponent.text = b64encode(long_to_bytes(key.public_key().public_numbers().e)).decode()
         elif self.sign_alg.name.startswith("DSA_"):
-            dsa_key_value = SubElement(key_value, ds_tag("DSAKeyValue"))
+            dsa_key_value = SubElement(key_value, self._ds_tag("DSAKeyValue"))
             for field in "p", "q", "g", "y":
-                e = SubElement(dsa_key_value, ds_tag(field.upper()))
+                e = SubElement(dsa_key_value, self._ds_tag(field.upper()))
 
                 if field == "y":
                     key_params = key.public_key().public_numbers()

--- a/test/test.py
+++ b/test/test.py
@@ -479,6 +479,33 @@ class TestSignXML(unittest.TestCase, LoadExampleKeys):
         expected_match = f'<Signature xmlns="{namespaces["ds"]}">'
         self.assertTrue(re.search(expected_match.encode("ascii"), signed_data))
 
+    def test_default_namespace_c14n_no_xmlns_undeclarations(self):
+        """
+        Test that using default namespace doesn't produce xmlns="" undeclarations in C14N.
+
+        When signer.namespaces = {None: ds_namespace}, the SignedInfo C14N should not
+        contain xmlns="" on child elements. This regression was reported in issue #275.
+        """
+        data = etree.parse(self.example_xml_files[0]).getroot()
+        signer = XMLSigner()
+        signer.namespaces = {None: namespaces["ds"]}
+        signed = signer.sign(data, key=self.keys["rsa"])
+
+        # Verify signature round-trips correctly
+        XMLVerifier().verify(signed, x509_cert=self.certs["example"])
+
+        # Extract SignedInfo and canonicalize it
+        signed_info = signed.find(".//{http://www.w3.org/2000/09/xmldsig#}SignedInfo")
+        c14n_output = etree.tostring(signed_info, method="c14n").decode()
+
+        # SignedInfo should have xmlns declaration, but child elements should NOT have xmlns=""
+        # Count occurrences: should be exactly 1 (on SignedInfo itself)
+        xmlns_count = c14n_output.count('xmlns="')
+        self.assertEqual(xmlns_count, 1, f"Expected 1 xmlns declaration, found {xmlns_count}. C14N output: {c14n_output}")
+
+        # Specifically verify no xmlns="" undeclarations
+        self.assertNotIn('xmlns=""', c14n_output, f"Found xmlns='' undeclaration in C14N output: {c14n_output}")
+
     def test_elementtree_compat(self):
         data = stdlibElementTree.parse(self.example_xml_files[0]).getroot()
         signer = XMLSigner()

